tgui AI Airlock

ThePotato97 opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->

[FEATURE] [UI]

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

![image](https://user-images.githubusercontent.com/10260415/91620980-728f3280-e989-11ea-8e34-49eabb5245dc.png)

![image](https://user-images.githubusercontent.com/10260415/91620992-76bb5000-e989-11ea-9831-def7d7051671.png)


Adds a new state in silicon.dm called tgui_silicon_state.
Adds AiAirlock.js which contains the frontend for the tgui
Edits airlock.dm splitting the old UI's functions into separate ones and commented out the old UI.

## Why's this needed? <!-- Describe why you think this should be added to the game. -->



## Changelog
<!-- If necessary, put your changelog entry below. Otherwise, please delete it.
Use however you want to be credited in the changelog in place of CodeDude.
Use (*) for major changes and (+) for minor changes. For example: -->

```
(u)ThePotato97:
(+)Converted AI airlock controls to tgui
```

Sovexe commented on code/obj/machinery/door/airlock.dm:
```diff
@@ -28,6 +28,174 @@
+/obj/machinery/door/airlock/ui_act(action, params)
+	if(..())
+		return
+	if(!src.allowed(usr))
+		return
+	if(!src.arePowerSystemsOn())
+		return
```
I feel like this is a weird place to check these

I'd expect the UI status to change, or the UI to have never opened in the first place if I couldn't access a door either due to lack or loss of access. Just blocking the users inputs leaves it unclear as to why they can't do anything

PrimeNumb replied:
In addition, perhaps more clear visual feedback for when you've lost/lack access? Just an idea though.

ThePotato97 replied:
I already disable the buttons when the PowerSystems aren't on this only checks if secondsMainPowerLost and 
secondsBackupPowerLost are 0 if the airlock loses power externally the eye will turn red and i'll have the UI update to show there's no external power.

Sovexe commented on code/modules/tgui/states/silicon.dm:
```diff
@@ -0,0 +1,12 @@
+
+/datum/ui_state/tgui_silicon_state/can_use_topic(src_object, mob/user)
+	if((issilicon(user) && !isghostdrone(user)) || isAIeye(user))
+		return UI_INTERACTIVE
```
This gives borgs infinite range which is inconsistent with the other states. The other borg related states require them to be within a certain distance to determine if they can open the UI, if the UI stays open, and if they can interact with the UI. Personally I think we should maintain that behavior here as well

This is also written very differently from the other states which are done by having the datum call the individual procs for each mob type

PrimeNumb replied:
I'd like to warn that removing cyborgs' ability to access doors from a range is a fairly major gameplay change and should probably not be part of a UI overhaul PR.

Sovexe replied:
Thinking on this a little more, I've changed my position. Having a check that does simply checks if you are a silicon should be fine. The behavior of allowing borgs access at range can still be achieved by combining with other checks such as `tgui_default_state`

That said this requires silicon state to be broken out into the individual mob procs still to return `UI_CLOSE` when mobs attempt to use it

Sovexe commented on tgui/packages/tgui/interfaces/AiAirlock.js:
```diff
@@ -0,0 +1,160 @@
+                  content={data.id_scanner ? 'Enabled' : 'Disabled'}
+                  selected={data.id_scanner}
+                  disabled={!data.wires.id_scanner
+                    || !data.power.main && !data.power.backup}
```
```suggestion
                    || (!data.power.main && !data.power.backup)}
```

Sovexe commented on tgui/packages/tgui/interfaces/AiAirlock.js:
```diff
@@ -0,0 +1,160 @@
+                  content={data.locked ? 'Lowered' : 'Raised'}
+                  selected={data.locked}
+                  disabled={!data.wires.bolts
+                    || !data.power.main && !data.power.backup}
```
```suggestion
                    || (!data.power.main && !data.power.backup)}
```

Sovexe commented on tgui/packages/tgui/interfaces/AiAirlock.js:
```diff
@@ -0,0 +1,160 @@
+                  content={data.opened ? 'Open' : 'Closed'}
+                  selected={data.opened}
+                  disabled={(data.locked || data.welded)
+                    || !data.power.main && !data.power.backup}
```
```suggestion
                    || (!data.power.main && !data.power.backup)}
```

Sovexe commented on code/obj/machinery/door/airlock.dm:
```diff
@@ -28,6 +28,174 @@
+	var/list/data = list()
+
+	var/list/power = list()
+	power["main"] = secondsMainPowerLost ? 0 : 2 // boolean
+	power["main_timeleft"] = secondsMainPowerLost
+	power["backup"] = secondsBackupPowerLost ? 0 : 2 // boolean
+	power["backup_timeleft"] = secondsBackupPowerLost
+	data["power"] = power
+
+	data["shock"] = secondsElectrified == 0 ? 2 : 0
```
Whats the point of these checks? Check if it's a non zero in JS
```suggestion
	power["main"] = secondsMainPowerLost // boolean
	power["main_timeleft"] = secondsMainPowerLost
	power["backup"] = secondsBackupPowerLost // boolean
	power["backup_timeleft"] = secondsBackupPowerLost
	data["power"] = power

	data["shock"] = !secondsElectrified
```

Sovexe commented on tgui/packages/tgui/interfaces/AiAirlock.js:
```diff
@@ -0,0 +1,160 @@
+import { Window } from '../layouts';
+
+const dangerMap = {
+  2: {
+    color: 'good',
+    localStatusText: 'Offline',
+  },
+  1: {
+    color: 'average',
+    localStatusText: 'Caution',
+  },
```
You never use the caution status

```suggestion
  1: {
    color: 'good',
    localStatusText: 'Offline',
  },
```

Sovexe commented:
Looking at this again - you didn't remove the code for the old menu. They will open both the old one and TGUI one with how this currently. The old unused menu code as well as calls and updates to it should all be stripped out

ThePotato97 commented:
![image](https://user-images.githubusercontent.com/10260415/91796007-0f1a3480-ec17-11ea-955e-3cc5a8244f8d.png)

Added this danger zone area and moved electrify to it

mordent-goonstation reviewed (changes requested):
Commenting on UI appearance, not looked at code yet. Consider these actions items, please request a review from me when implemented (or discussed and argued about or whatever) and I'll take a look at the implementation.

I think the window is far too wide. I realize that the "electrify" options make it so. Given that they have their own special section now, consider making them cover two lines then bringing the width way in.

The dividers between the ID Scan / Door Bolts / Door Control seem unnecessary. They may appear necessary now to couple the labels with the buttons as they're so far apart, but you can lose the dividers once you've brought the width it.

Icon usage: Not a fan of the lightbulb for disrupting main/backup power. Consider `plug` if we have it, maybe? Consider `door-open`/`door-closed` for the door open/close ones.

Random green button for ID scan being enabled seems unneeded when the rest aren't green.

For the assorted timers, consider using the way they were implemented in [this PR](https://github.com/goonstation/goonstation/pull/1802) (not the visual style, but the way they try to update in real-seconds) if you haven't already.

ThePotato97 commented:
New UI changes:
![image](https://user-images.githubusercontent.com/10260415/92192625-50e8ec00-ee5e-11ea-8479-60d625c3140c.png)
![image](https://user-images.githubusercontent.com/10260415/92192666-65c57f80-ee5e-11ea-8df2-1ab155c029c9.png)
![image](https://user-images.githubusercontent.com/10260415/92192676-6b22ca00-ee5e-11ea-92f6-18bcda7f69a8.png)

ZeWaka commented on code/modules/tgui/states/silicon.dm:
```diff
@@ -0,0 +1,61 @@
+ *
+ * Checks that the user is an AI/borg.
+ */
+var/global/datum/ui_state/tgui_silicon_state/tgui_silicon_state = new /datum/ui_state/tgui_silicon_state
```
looks fine for the most part, this is a bit weird:
```
	if(. == UI_INTERACTIVE) // Non-human living mobs can only look, not touch.
		return UI_UPDATE
```

stylemistake replied:
Looks fine, but I'd love to see ui_state to just be gone in general from tgui.

mordent-goonstation commented on tgui/packages/tgui/interfaces/Airlock.js:
```diff
@@ -1,8 +1,417 @@
+
+  return (
+    <Box
+      m={-1} py={0.5} pt={2}
```
`-1` margin seems weird. Is this to make it so that it sits flush within the box/section above which normally has padding? Could you add the `fitted` prop to that container and apply appropriate margin to the other children? I would prefer applying margin a few times than applying negative margin once, as it's easier to understand.

mordent-goonstation commented on tgui/packages/tgui/interfaces/Airlock.js:
```diff
@@ -1,8 +1,417 @@
+    welded,
+  } = data;
+
+  const isDisabled = (data.status === 2);
```
While these may be "magic numbers" definitely have a variable that says what each of these is somewhere in here.

mordent-goonstation commented on tgui/packages/tgui/interfaces/Airlock.js:
```diff
@@ -1,8 +1,417 @@
+          )}>
+          {!!(!boltsAreUp || welded) && (
+            <span>
+              [Door is {!boltsAreUp ? "bolted" : ""}
+              {(!boltsAreUp && welded) ? " and " : ""}
+              {welded ? "welded" : ""}!]
```
Possible to do something like `[Door is {!boltsAreUp && 'bolted'}` etc. rather than ternaries here?

mordent-goonstation commented on tgui/packages/tgui/interfaces/Airlock.js:
```diff
@@ -1,8 +1,417 @@
+  return {
+    airlock: (userStates.isBorg) || (userStates.isAi),
+    accessPanel: (
+      (userStates.isBorg && userStates.distance <= 3 && panelOpen)
```
Making sure this wasn't missed. What's the `<= 3` check for?

ThePotato97 commented on tgui/packages/tgui/interfaces/Airlock.js:
```diff
@@ -1,8 +1,419 @@
+        ? 1
+        : userPerms["accessPanel"]
+          ? 2
+          : 1);
```
We no longer need this so i just made it all 1 cause we just show the user individual windows with what they need to see airlock, airlock+accesspanel, accesspanel.

ThePotato97 commented:
> Making sure this wasn't missed. What's the <= 3 check for?

It was to extend the range for borgs so they wouldn't switch to the airlock UI the second they get pushed or something, but I decided it wasn't necessary.
